File: recce/util/cloud/check_events.py
"""
Recce Cloud API client for check event operations.

This module provides methods for managing check events (timeline/conversation) in Recce Cloud,
including CRUD operations for comments and retrieving state change events.
"""

from typing import Dict, List

from recce.util.cloud.base import CloudBase


class CheckEventsCloud(CloudBase):
    """
    Client for Recce Cloud check event operations.

    Provides methods to list, create, update, and delete check events
    (comments and state changes) within Recce Cloud sessions.

    Examples:
        >>> client = CheckEventsCloud(token="your-api-token")
        >>> events = client.list_events(
        ...     org_id="org123",
        ...     project_id="proj456",
        ...     session_id="sess789",
        ...     check_id="check001"
        ... )
    """

    def _build_events_url(self, org_id: str, project_id: str, session_id: str, check_id: str) -> str:
        """Build the base URL for check events endpoints."""
        return f"{self.base_url_v2}/organizations/{org_id}/projects/{project_id}/sessions/{session_id}/checks/{check_id}/events"

    def list_events(self, org_id: str, project_id: str, session_id: str, check_id: str) -> List[Dict]:
        """
        List all events for a check in chronological order.

        Args:
            org_id: Organization ID
            project_id: Project ID
            session_id: Session ID
            check_id: Check ID

        Returns:
            List of event dictionaries

        Raises:
            RecceCloudException: If the request fails

        Example:
            >>> events = client.list_events("org123", "proj456", "sess789", "check001")
            >>> for event in events:
            ...     print(f"{event['event_type']}: {event['content']}")
        """
        api_url = self._build_events_url(org_id, project_id, session_id, check_id)
        query_params = {"include_deleted": True}
        response = self._request("GET", api_url, params=query_params)

        self._raise_for_status(
            response,
            "Failed to list check events from Recce Cloud.",
        )

        data = response.json()
        # Response is wrapped: {"events": [...]}
        return data.get("events", [])

    def get_event(self, org_id: str, project_id: str, session_id: str, check_id: str, event_id: str) -> Dict:
        """
        Get a specific event by ID.

        Args:
            org_id: Organization ID
            project_id: Project ID
            session_id: Session ID
            check_id: Check ID
            event_id: Event ID

        Returns:
            Event dictionary

        Raises:
            RecceCloudException: If the request fails or event not found
        """
        api_url = f"{self._build_events_url(org_id, project_id, session_id, check_id)}/{event_id}"
        response = self._request("GET", api_url)

        self._raise_for_status(
            response,
            f"Failed to get check event {event_id} from Recce Cloud.",
        )

        data = response.json()
        # Response is wrapped: {"event": {...}}
        return data.get("event", {})

    def create_comment(self, org_id: str, project_id: str, session_id: str, check_id: str, content: str) -> Dict:
        """
        Create a new comment on a check.

        Args:
            org_id: Organization ID
            project_id: Project ID
            session_id: Session ID
            check_id: Check ID
            content: Comment content (plain text or markdown)

        Returns:
            Created event dictionary

        Raises:
            RecceCloudException: If the request fails

        Example:
            >>> event = client.create_comment(
            ...     "org123", "proj456", "sess789", "check001",
            ...     "This looks good to me!"
            ... )
            >>> print(f"Created comment with ID: {event['id']}")
        """
        api_url = self._build_events_url(org_id, project_id, session_id, check_id)
        response = self._request("POST", api_url, json={"content": content})

        self._raise_for_status(
            response,
            "Failed to create comment in Recce Cloud.",
        )

        data = response.json()
        # Response is wrapped: {"event": {...}}
        return data.get("event", {})

    def update_comment(
        self, org_id: str, project_id: str, session_id: str, check_id: str, event_id: str, content: str
    ) -> Dict:
        """
        Update an existing comment.

        Only the author or an admin can update a comment.

        Args:
            org_id: Organization ID
            project_id: Project ID
            session_id: Session ID
            check_id: Check ID
            event_id: Event ID of the comment to update
            content: New comment content

        Returns:
            Updated event dictionary

        Raises:
            RecceCloudException: If the request fails or user is not authorized
        """
        api_url = f"{self._build_events_url(org_id, project_id, session_id, check_id)}/{event_id}"
        response = self._request("PATCH", api_url, json={"content": content})

        self._raise_for_status(
            response,
            f"Failed to update comment {event_id} in Recce Cloud.",
        )

        data = response.json()
        # Response is wrapped: {"event": {...}}
        return data.get("event", {})

    def delete_comment(self, org_id: str, project_id: str, session_id: str, check_id: str, event_id: str) -> None:
        """
        Delete a comment (soft delete).

        Only the author or an admin can delete a comment.
        The comment will be marked as deleted but remain in the timeline.

        Args:
            org_id: Organization ID
            project_id: Project ID
            session_id: Session ID
            check_id: Check ID
            event_id: Event ID of the comment to delete

        Raises:
            RecceCloudException: If the request fails or user is not authorized
        """
        api_url = f"{self._build_events_url(org_id, project_id, session_id, check_id)}/{event_id}"
        response = self._request("DELETE", api_url)

        self._raise_for_status(
            response,
            f"Failed to delete comment {event_id} in Recce Cloud.",
        )

File: js/packages/ui/src/api/checkEvents.ts
/**
 * Check Events API - Types and Client Functions
 *
 * Provides timeline/conversation functionality for Checks, enabling
 * GitHub PR-style discussions with comments and automatic tracking
 * of state changes.
 *
 * NOTE: This feature is only available when connected to Recce Cloud.
 */

import type { AxiosInstance, AxiosResponse } from "axios";

// ============================================================================
// Event Types
// ============================================================================

export type CheckEventType =
  | "check_created"
  | "comment"
  | "approval_change"
  | "description_change"
  | "name_change"
  | "preset_applied";

export type ActorType = "user" | "recce_ai" | "preset_system";

// ============================================================================
// Actor Interface
// ============================================================================

export interface CheckEventActor {
  type: ActorType;
  user_id: number | null;
  login: string | null;
  fullname: string | null;
}

// ============================================================================
// Check Event Interface
// ============================================================================

export interface CheckEvent {
  id: string;
  check_id: string;
  event_type: CheckEventType;
  actor: CheckEventActor;
  content: string | null;
  old_value: string | null;
  new_value: string | null;
  is_edited: boolean;
  is_deleted: boolean;
  created_at: string;
  updated_at: string;
}

// ============================================================================
// Request/Response Types
// ============================================================================

export interface CreateCommentRequest {
  content: string;
}

export interface UpdateCommentRequest {
  content: string;
}

export interface CheckEventsListResponse {
  events: CheckEvent[];
}

// ============================================================================
// API Client Functions
// ============================================================================

/**
 * List all events for a check in chronological order.
 *
 * @param checkId - The check ID
 * @param client - Axios instance for API configuration
 * @returns Promise resolving to array of CheckEvent objects
 */
export async function listCheckEvents(
  checkId: string,
  client: AxiosInstance,
): Promise<CheckEvent[]> {
  const response = await client.get<never, AxiosResponse<CheckEvent[]>>(
    `/api/checks/${checkId}/events`,
  );
  return response.data;
}

/**
 * Get a specific event by ID.
 *
 * @param checkId - The check ID
 * @param eventId - The event ID
 * @param client - Axios instance for API configuration
 * @returns Promise resolving to a CheckEvent object
 */
export async function getCheckEvent(
  checkId: string,
  eventId: string,
  client: AxiosInstance,
): Promise<CheckEvent> {
  const response = await client.get<never, AxiosResponse<CheckEvent>>(
    `/api/checks/${checkId}/events/${eventId}`,
  );
  return response.data;
}

/**
 * Create a new comment on a check.
 *
 * @param checkId - The check ID
 * @param content - The comment content (plain text for now, markdown later)
 * @param client - Axios instance for API configuration
 * @returns Promise resolving to the created CheckEvent
 */
export async function createComment(
  checkId: string,
  content: string,
  client: AxiosInstance,
): Promise<CheckEvent> {
  const response = await client.post<
    CreateCommentRequest,
    AxiosResponse<CheckEvent>
  >(`/api/checks/${checkId}/events`, { content });
  return response.data;
}

/**
 * Update an existing comment.
 * Only the author or an admin can update a comment.
 *
 * @param checkId - The check ID
 * @param eventId - The event ID of the comment to update
 * @param content - The new comment content
 * @param client - Axios instance for API configuration
 * @returns Promise resolving to the updated CheckEvent
 */
export async function updateComment(
  checkId: string,
  eventId: string,
  content: string,
  client: AxiosInstance,
): Promise<CheckEvent> {
  const response = await client.patch<
    UpdateCommentRequest,
    AxiosResponse<CheckEvent>
  >(`/api/checks/${checkId}/events/${eventId}`, { content });
  return response.data;
}

/**
 * Delete a comment (soft delete).
 * Only the author or an admin can delete a comment.
 *
 * @param checkId - The check ID
 * @param eventId - The event ID of the comment to delete
 * @param client - Axios instance for API configuration
 * @returns Promise resolving when deletion is complete
 */
export async function deleteComment(
  checkId: string,
  eventId: string,
  client: AxiosInstance,
): Promise<void> {
  await client.delete(`/api/checks/${checkId}/events/${eventId}`);
}

// ============================================================================
// Type Guards
// ============================================================================

export function isCommentEvent(event: CheckEvent): boolean {
  return event.event_type === "comment";
}

export function isStateChangeEvent(event: CheckEvent): boolean {
  return (
    event.event_type === "approval_change" ||
    event.event_type === "description_change" ||
    event.event_type === "name_change"
  );
}

export function isSystemEvent(event: CheckEvent): boolean {
  return (
    event.event_type === "check_created" ||
    event.event_type === "preset_applied"
  );
}

// ============================================================================
// Display Helpers
// ============================================================================

/**
 * Get a human-readable description of an event for display.
 */
export function getEventDescription(event: CheckEvent): string {
  const actorName = event.actor.fullname || event.actor.login || "Someone";

  switch (event.event_type) {
    case "check_created":
      return `${actorName} created this check`;
    case "comment":
      if (event.is_deleted) {
        return "Comment deleted";
      }
      return event.content || "";
    case "approval_change":
      if (event.new_value === "true") {
        return `${actorName} approved this check`;
      }
      return `${actorName} unapproved this check`;
    case "description_change":
      return `${actorName} updated the description`;
    case "name_change":
      return `${actorName} renamed this check`;
    case "preset_applied":
      return `${actorName} applied a preset`;
    default:
      return `${actorName} made a change`;
  }
}

/**
 * Get the appropriate icon name for an event type.
 * Returns a string that can be used with react-icons.
 */
export function getEventIconType(
  event: CheckEvent,
): "comment" | "approve" | "unapprove" | "edit" | "create" | "preset" {
  switch (event.event_type) {
    case "check_created":
      return "create";
    case "comment":
      return "comment";
    case "approval_change":
      return event.new_value === "true" ? "approve" : "unapprove";
    case "description_change":
    case "name_change":
      return "edit";
    case "preset_applied":
      return "preset";
    default:
      return "edit";
  }
}
